File: 278-first-bad-version/278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        # use binary search
        first_bad = n
        
        lo, hi = 1, n
        
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if (isBadVersion(mid)):
                if mid < first_bad:
                    first_bad = mid # update first bad
                hi = mid # move threshold down
            else: # midpoint is OK
                lo = mid + 1
        return first_bad

        

File: 278-first-bad-version/278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        // use 2 ptrs in binary search
        int lo = 1;
        int hi = n;
        int first_bad = n;
        
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (isBadVersion(mid)) {
                if (mid < first_bad) {
                    first_bad = mid;
                }
                hi = mid;
            }
            else {
                lo = mid + 1;
            }
        }
        
        return first_bad;
    }
};